### ISSUE TITLE
pass sf::Vector2<T> and sf::Vector3<T> by value instead of by references

### ISSUE THREAD

**Issue by ZXShady** [SFML/SFML#3047]
### Prerequisite Checklist

- [X] I searched for [existing issues](https://github.com/search?q=repo%3ASFML%2FSFML&type=issues) to prevent duplicates
- [X] I searched for [existing discussions on the forum](https://www.google.com/search?q=site%3Ahttps%3A%2F%2Fen.sfml-dev.org) to prevent duplicates
- [X] I understand [SFML's scope](https://www.sfml-dev.org/contribute.php#general-considerations) and believe the feature fits it

### Describe your feature request here

Simple structs like these should be passed by value instead of by references in many places and get* functions should return by value too.

### Use Cases

it is just for extra performance for functions defined out of line.
as the that function could internally use a `const_cast`  on the reference passed legally that would disable many optimizations and the compiler had to take into account aliasing issues with global variables.

Simple trivially copyable structs/classes should be passed by value to avoid rhe issue above.

 functions that are defined in another translation unit and takes a parameter by const reference the caller has to assume that the called function could change the paramter and the compiler has to take that into account
`void f(int const&)` f could change its arguement and that is legal only if the passed int arguement was not orignally const.

**Comment by vittorioromeo** [comment]
Can you provide a benchmark where this matters? 

Generally speaking, if the function is called so frequently that it becomes significant in the benchmark, it should be inlined and whether the vector is passed by value or by reference shouldn't be important.

**Comment by ChrisThrasher** [comment]
As far as I’m aware there’s no noticeable difference between passing a `sf::Vector2i` or `sf::Vector2f` by value or const reference. However I do think 3D vectors are large enough that passing by const reference is superior so I don’t see why we’d ever change 3D vector parameter to be passed by value.

**Comment by ChrisThrasher** [comment]
Closing this issue until we receive more evidence for why passing by value is an improvement.

**Comment by ZXShady** [comment]
hey sorry for the extremely late reply yikes 3 whole weeks.

passing by reference means if the compiler cannot know the body of the function means that your variable could be theoritcally changed by the function


```
void print(sf::Vector2f const&); // defined in another translation unit

int main() {
    sf::Vector2f v{4,2};
    print(v);
    return v.x; // won't be optimized into return 4 since this could change by the function
// since the function could const_cast it away!
}

```
[by reference](https://godbolt.org/z/axTbo8hMo)
[by value](https://godbolt.org/z/zEPx47T8s )

**Comment by ChrisThrasher** [comment]
Reopened because you provide some more evidence. Haven't looked deeply into it yet but it seems worth consideration.

**Comment by ChrisThrasher** [comment]
Curious what @vittorioromeo has to say about passing vectors by value having optimizer implications. This is not an argument I've heard before with regard to value vs const ref parameters, but those godbolt links seem to illustrate it well.

**Comment by ZXShady** [comment]
It also saves keystrokes I say that is another benefit /joke

sad that const references are not really `const` if you get what I mean.

**Comment by ZXShady** [comment]
> Can you provide a benchmark where this matters? 
> 
> Generally speaking, if the function is called so frequently that it becomes significant in the benchmark, it should be inlined and whether the vector is passed by value or by reference shouldn't be important.
> 
> I'm also not convinced that passing something like `Vector3` by value would be faster than by-reference in all cases, again a benchmark would help.

definitely Vector3 of float and int would be faster as for double it may not be faster or slower but without any testing but I think it would be faster still due to guranteeing that the function is not modifying the variable.


functions defined in another translation unit (most of SFML function are defined in another translation unit)

then it cannot be inlined without LTO

**Comment by ZXShady** [comment]
> Can you provide a benchmark where this matters? 
> 
> Generally speaking, if the function is called so frequently that it becomes significant in the benchmark, it should be inlined and whether the vector is passed by value or by reference shouldn't be important.
> 
> I'm also not convinced that passing something like `Vector3` by value would be faster than by-reference in all cases, again a benchmark would help.

definitely Vector3 of float and int would be faster as for double it may not be faster or slower but without any testing but I think it would be faster still due to guranteeing that the function is not modifying the variable.


functions defined in another translation unit (most of SFML function are defined in another translation unit)

then it cannot be inlined without LTO

**Comment by ChrisThrasher** [comment]
I'm of the opinion that a change of this magnitude requires some empirical evidence. What you're saying sounds plausible but I'm reluctant to make such a broad, sweeping change unless I'm confident it results in better code generation by the compiler.

**Comment by ZXShady** [comment]
@ChrisThrasher 

What more evidence do you need (asking politly)?

most SFML functions are not defined `inline` and taking const reference parameters means that the variable could be changed inside the function by const_cast as representable by the godbolt links (simplified) and this is 100% legal and the standard obligates the compiler to not assume about the functions, the example is simple but they represent the piont out of line functions taking const lvalues could mosify their parameters legally 

```cpp
void f(const int& x,bool is_const) {
     if(!is_const) 
           const_cast<int&>(x) = 5; // legal if x is not referencing a const object
}

//code
int i = 0;
f(i,false);
std::cout << i; // the compiler is not allowed to replace it by 0 since the function if cannot be inlined can modify the value.

const int ci = 0;
f(ci,true);
std::cout << ci; // const objects cannot be modified and modifying them by const cast results in undefined behavior.
```

Now is this code good lol no but the compiler has to take this in account and all compilers msvc,clang,gcc all take this into account. and most SFML functions don't need to reference the value they just need the values in there so take them by copy and make the optimizer more happy.

also I would recommend making making one line functions that are bassicly wrapper as inline to prevent kinda costly mistakes like this

```cpp
sf::VertexBuffer v;
for(std::size_t i = 0;i < v.size();++i); // oops! calling v.size() everytime it is defined out of line so that means it will call it everytime!
```

**Comment by ZXShady** [comment]
If you want stronger evidence

you can try asking yourself why are int and floats and doubles passed by value and not by const reference? apply that to these simple structures who are nothing more than a wrapper around them

**Comment by Bambo-Borris** [comment]
> @ChrisThrasher 
> 
> What more evidence do you need (asking politly)?
> 

As mentioned previously, I believe Thrasher is seeking some benchmarks to give empirical evidence of the speed up gains you mentioned. I think it's a reasonable request in this scenario and I think others would also be interested in seeing the results.

**Comment by ChrisThrasher** [comment]
Here are the rules I was working with as of yesterday to determine whether a parameter ought to be a value or `const&`:

1. `sizeof(T)` < CPU register width -> value parameter
2. `sizeof(T)` == CPU register width -> doesn't matter
3. `sizeof(T)` > CPU register width -> `const&` parameter

I've discussed this more with some C++ peers of mine and have come to learn some details I wasn't previously aware of.  I now realize my mental model is incorrect. That `in<T>` template you shared matches with what they described with the small difference of using `<=` instead of `<` as you wrote. The idea of using a template to determine parameter type is really interesting. It's probably overkill for SFML but I think it's a really interesting idea.

The most common 2D vectors are `sf::Vector2i`, `sf::Vector2u` and `sf::Vector2f`. All 3 of these template specializations use template arguments that are 32 bits wide. That means all three of these types fit into the `<= 2 * CPU register` criteria whether we're talking about 32 bit or 64 bit platforms. For that reason I think it's a good idea to convert those to value parameters like you're proposing. I think this is a change we can justify based on first principles without bothering with benchmarks.

3D vectors are a bit harder since common specialization like `Vector3f` are bigger than 2 CPU registers on 32 bit platforms but less than 2 CPU registers on 64 bit platforms. The C++ Core Guidelines guideline you referenced says "two or three words (doubles, pointers, references) are usually best passed by value" so perhaps this is another case were we use value parameters knowing it's good enough on 32 bits and optimal on 64 bit platforms. While we treat 32 bit platforms as a first class citizen, ultimately we ought to prefer 64 bit compiles since those are much more common and will only be more common at times goes on.

TL;DR I think this is a good idea 👍🏻

**Comment by ChrisThrasher** [comment]
https://github.com/SFML/SFML/compare/master...value_parameters

Did some quick find/replace to swap over to value parameters to see what it would look like.

**Comment by ChrisThrasher** [comment]
https://github.com/SFML/SFML/pull/3140

At a minimum we can make this change for all 2D vectors. Whether we make the change for 3D vectors can be discussed further and handled in a separate PR if at all. Certainly the majority of the value to be gained is by changing how we handle 2D vectors which are used much more prolifically than 3D vectors.

**Comment by ZXShady** [comment]
>sizeof(T) < CPU register width -> value parameter
>sizeof(T) == CPU register width -> doesn't matter
>sizeof(T) > CPU register width -> const& parameter


well yea this is bad that means if using a 32 bit platform you will pass double by const reference.

**Comment by ZXShady** [comment]
> https://github.com/SFML/SFML/pull/3140
> 
> At a minimum we can make this change for all 2D vectors. Whether we make the change for 3D vectors can be discussed further and handled in a separate PR if at all. Certainly the majority of the value to be gained is by changing how we handle 2D vectors which are used much more prolifically than 3D vectors.



for 3D vectors of float and int they are still worth it to be passed by value because inside the functions they have to do pointer derefencing while by value is not.

the core guidelines say 2 or 3 words I took the minimum to be absolutely safe.

**Comment by ZXShady** [comment]
@Bambo-Borris 

I have no benchmark ready right now But I have posted Godbolt links that show the differences and optimizations possible by pass by value

**Comment by ChrisThrasher** [comment]
Closed because I think we're done as much towards this goal as is going to happen. Future PRs welcome if any low hanging fruit from missed, but for the most part we've now addressed this with #3140, #3170, and #3179.
